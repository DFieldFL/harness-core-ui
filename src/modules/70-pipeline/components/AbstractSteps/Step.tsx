import type { IconName } from '@wings-software/uikit'
import type { CompletionItemInterface } from '@common/interfaces/YAMLBuilderProps'

export enum StepViewType {
  InputSet = 'InputSet',
  InputVariable = 'InputVariable',
  DeploymentForm = 'DeploymentForm',
  StageVariable = 'StageVariable',
  Edit = 'Edit'
}
export abstract class Step<T extends object> {
  protected abstract type: string
  protected abstract defaultValues: T
  protected abstract stepIcon: IconName
  protected abstract stepName: string
  protected invocationMap?: Map<RegExp, (path: string, yaml: string) => Promise<CompletionItemInterface[]>>

  protected stepPaletteVisible?: boolean // default to true

  getType(): string {
    return this.type
  }

  getDefaultValues(initialValues: T, _stepViewType: StepViewType): T {
    return { ...this.defaultValues, ...initialValues }
  }

  getIconName(): IconName {
    return this.stepIcon
  }

  getStepName(): string {
    return this.stepName
  }

  getInvocationMap(): Map<RegExp, (path: string, yaml: string) => Promise<CompletionItemInterface[]>> | undefined {
    return this.invocationMap
  }

  getStepPaletteVisibility(): boolean {
    return this.stepPaletteVisible ?? true
  }

  abstract renderStep(
    initialValues: T,
    onUpdate?: (data: T) => void,
    stepViewType?: StepViewType,
    template?: { [P in keyof T]: string },
    allValues?: T
  ): JSX.Element
}
