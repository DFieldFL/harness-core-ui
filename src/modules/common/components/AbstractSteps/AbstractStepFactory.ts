import type { IconName } from '@wings-software/uikit'
import type { Step } from './Step'

export interface StepData {
  label: string
  icon: IconName
  type: string
  visible: boolean
}

export abstract class AbstractStepFactory {
  /**
   * Couples the factory with the steps it generates
   */
  protected abstract type: string

  protected stepBank: Map<string, Step<object>>
  protected stepIconMap: Map<string, StepData>

  constructor() {
    this.stepBank = new Map()
    this.stepIconMap = new Map()
  }

  getType(): string {
    return this.type
  }

  registerStep<T extends object>(step: Step<T>): void {
    this.stepBank.set(step.getType(), step)
    this.stepIconMap.set(step.getType(), {
      label: step.getStepName(),
      icon: step.getIconName(),
      type: step.getType(),
      visible: step.getStepPaletteVisibility()
    })
  }

  deregisterStep(type: string): void {
    this.stepBank.delete(type)
    this.stepIconMap.delete(type)
  }

  getStep<T extends object>(type: string): Step<T> | undefined {
    return this.stepBank.get(type) as Step<T>
  }

  getStepIcon(type: string): IconName | undefined {
    return this.stepBank.get(type)?.getIconName()
  }

  getStepData(type: string): StepData | undefined {
    return this.stepIconMap.get(type)
  }

  getAllStepsDataList(): Array<StepData> {
    return Array.from(this.stepIconMap, ([_key, value]) => value).filter(step => step.visible)
  }
}
