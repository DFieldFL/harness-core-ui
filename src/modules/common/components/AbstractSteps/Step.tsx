import type { IconName } from '@wings-software/uikit'

export enum StepViewType {
  InputSet = 'InputSet',
  DeploymentForm = 'DeploymentForm',
  Edit = 'Edit'
}
export abstract class Step<T extends object> {
  protected abstract type: string
  protected abstract defaultValues: T
  protected abstract stepIcon: IconName
  protected abstract stepName: string

  getType(): string {
    return this.type
  }

  getDefaultValues(): T {
    return this.defaultValues
  }

  getIconName(): IconName {
    return this.stepIcon
  }

  getStepName(): string {
    return this.stepName
  }

  abstract renderStep(initialValues: T, onUpdate?: (data: T) => void, stepViewType?: StepViewType): JSX.Element
}
