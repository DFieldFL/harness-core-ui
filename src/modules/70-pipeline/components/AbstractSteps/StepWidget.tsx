import React from 'react'
import { Text } from '@wings-software/uikit'
import type { AbstractStepFactory } from './AbstractStepFactory'
import i18n from './StepWidget.i18n'
import { StepViewType } from './Step'

export interface StepWidgetProps<T extends object = {}> {
  factory: AbstractStepFactory
  type: string
  initialValues: T
  allValues?: T
  template?: T
  stepViewType?: StepViewType
  onUpdate?: (data: T) => void
}

export function StepWidget<T extends object = {}>({
  factory,
  type,
  initialValues,
  allValues,
  template,
  stepViewType = StepViewType.Edit,
  onUpdate
}: StepWidgetProps<T>): JSX.Element {
  const step = factory.getStep<T>(type)

  if (!step) {
    return <Text intent="warning">{i18n.invalidStep}</Text>
  } else {
    const values = step?.getDefaultValues(initialValues, stepViewType)
    return <>{step.renderStep(values, onUpdate, stepViewType, template, allValues)}</>
  }
}
