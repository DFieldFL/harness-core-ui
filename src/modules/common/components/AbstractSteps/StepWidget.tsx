import React from 'react'
import { Text } from '@wings-software/uikit'
import type { AbstractStepFactory } from './AbstractStepFactory'
import i18n from './StepWidget.i18n'
import { StepViewType } from './Step'

export interface StepWidgetProps {
  factory: AbstractStepFactory
  type: string
  initialValues: object
  stepViewType?: StepViewType
  onUpdate?: (data: object) => void
}

export function StepWidget({
  factory,
  type,
  initialValues,
  stepViewType = StepViewType.Edit,
  onUpdate
}: StepWidgetProps): JSX.Element {
  const step = factory.getStep(type)

  if (!step) {
    return <Text intent="warning">{i18n.invalidStep}</Text>
  } else {
    const values = { ...step?.getDefaultValues(), ...initialValues }
    return <>{step.renderStep(values, onUpdate, stepViewType)}</>
  }
}
