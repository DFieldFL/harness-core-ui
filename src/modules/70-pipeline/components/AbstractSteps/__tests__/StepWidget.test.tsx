/* eslint-disable @typescript-eslint/ban-types */
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import type { IconName } from '@wings-software/uicore'
import type { StepType } from '@pipeline/components/PipelineSteps/PipelineStepInterface'
import { AbstractStepFactory } from '../AbstractStepFactory'
import { Step, StepProps } from '../Step'
import { StepWidget } from '../StepWidget'

class StepFactory extends AbstractStepFactory {
  protected type = 'test-factory'
}

class StepOne extends Step<object> {
  protected type = 'step-one' as StepType
  protected stepName = 'stepOne'
  protected stepIcon: IconName = 'cross'
  validateInputSet(): object {
    return {}
  }
  protected defaultValues = { a: 'a' }
  renderStep(props: StepProps<object>): JSX.Element {
    return <div onClick={() => props.onUpdate?.(props.initialValues)}>{JSON.stringify(props.initialValues)}</div>
  }
}

class StepTwo extends Step<object> {
  protected type = 'step-two' as StepType
  protected stepName = 'stepTwo'
  protected stepIcon: IconName = 'cross'
  validateInputSet(): object {
    return {}
  }

  protected defaultValues = { b: 'b' }
  renderStep(props: StepProps<object>): JSX.Element {
    return <div onClick={() => props.onUpdate?.(props.initialValues)}>{JSON.stringify(props.initialValues)}</div>
  }
}

const factory = new StepFactory()
factory.registerStep(new StepOne())
factory.registerStep(new StepTwo())

describe('StepWidget tests', () => {
  test(`shows different steps based on type`, () => {
    let { container } = render(<StepWidget type={'step-one' as StepType} factory={factory} initialValues={{}} />)
    expect(container).toMatchSnapshot()
    container = render(<StepWidget type={'step-two' as StepType} factory={factory} initialValues={{}} />).container
    expect(container).toMatchSnapshot()
  })

  test(`shows invalid step`, () => {
    const { container } = render(<StepWidget type={'step-three' as StepType} factory={factory} initialValues={{}} />)
    expect(container).toMatchSnapshot()
  })

  test(`shows step and merge initial values`, () => {
    const { container } = render(
      <StepWidget type={'step-one' as StepType} factory={factory} initialValues={{ a: 'b' }} />
    )
    expect(container).toMatchSnapshot()
  })

  test(`should call on submit of the form`, () => {
    const onSubmit = jest.fn()
    const { container } = render(
      <StepWidget type={'step-one' as StepType} onUpdate={onSubmit} factory={factory} initialValues={{ a: 'b' }} />
    )
    fireEvent.click(container.children[0])
    expect(onSubmit).toBeCalled()
  })
})
