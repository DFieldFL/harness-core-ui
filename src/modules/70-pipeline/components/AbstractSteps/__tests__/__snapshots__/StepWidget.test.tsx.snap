// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepWidget tests shows different steps based on type 1`] = `
<div>
  <div>
    {"a":"a"}
  </div>
</div>
`;

exports[`StepWidget tests shows different steps based on type 2`] = `
<div>
  <div>
    {"b":"b"}
  </div>
</div>
`;

exports[`StepWidget tests shows invalid step 1`] = `
<div>
  <p
    class="StyledProps---font--2ipy StyledProps---main--1Wws StyledProps---intent-warning--2N1t"
  >
    Step not found
  </p>
</div>
`;

exports[`StepWidget tests shows step and merge initial values 1`] = `
<div>
  <div>
    {"a":"b"}
  </div>
</div>
`;
